cmake/therock_subproject.cmake

```
@@ -1429,3 +1537,86 @@ function(_therock_cmake_subproject_setup_toolchain
   set(_build_env_pairs "${_build_env_pairs}" PARENT_SCOPE)
   file(CONFIGURE OUTPUT "${toolchain_file}" CONTENT "${_toolchain_contents}" @ONLY ESCAPE_QUOTES)
 endfunction()
+
+
+# Fingerprints a source directory. If the source directory is considered dirty
+# then the output fingerprint will be the empty string (invalid).
+function(_therock_subproject_fprint_source_dir out_fprint dir)
+  set(fprint)
+  cmake_path(GET dir PARENT_PATH _source_parent)
+  cmake_path(GET dir FILENAME _source_basename)
+  set(_smrev_file "${_source_parent}/.${_source_basename}.smrev")
+
+  if(EXISTS "${_smrev_file}")
+    # The source control system has hard-coded a stable fingerprint that should
+    # be taken verbatim.
+    file(READ "${_smrev_file}" fprint)
+  endif()
+
+  # Determine if a git repo.
+  execute_process(
+      COMMAND git rev-parse --git-dir
+      WORKING_DIRECTORY "${dir}"
+      RESULT_VARIABLE GIT_CHECK_RESULT
+      OUTPUT_QUIET
+      ERROR_QUIET
+  )
+  if(GIT_CHECK_RESULT EQUAL 0)
+      set(IN_GIT_REPO TRUE)
+  else()
+      set(IN_GIT_REPO FALSE)
+  endif()
+
+  # Handle git repo.
+  if(IN_GIT_REPO)
+    if(NOT fprint)
+      # Process as git directory.
+      # Get the HEAD commit hash
+      execute_process(
+        COMMAND git rev-parse HEAD
+        WORKING_DIRECTORY "${dir}"
+        OUTPUT_VARIABLE GIT_COMMIT_HASH
+        OUTPUT_STRIP_TRAILING_WHITESPACE
+        RESULT_VARIABLE GIT_HASH_RESULT
+        ERROR_QUIET
+      )
+      if(GIT_HASH_RESULT EQUAL 0)
+        set(fprint "${GIT_COMMIT_HASH}")
+      else()
+        if(THEROCK_VERBOSE)
+          message(STATUS "  FPRINT: Could not read git commit")
+        endif()
+        set(fprint)
+      endif()
+    endif()
+
+    # Check if working tree is pristine (no modified tracked files)
+    execute_process(
+      COMMAND git diff --quiet HEAD
+      WORKING_DIRECTORY "${dir}"
+      RESULT_VARIABLE GIT_DIFF_RESULT
+    )
```

RE: the other comment thread, how about adding `--ignore-submodules` here?

https://git-scm.com/docs/git-diff#Documentation/git-diff.txt---ignore-submodulesnoneuntrackeddirtyall

```suggestion
    # Check if working tree is pristine (no modified tracked files)
    execute_process(
      COMMAND git diff --quiet --ignore-submodules HEAD
      WORKING_DIRECTORY "${dir}"
      RESULT_VARIABLE GIT_DIFF_RESULT
    )
```

---

Added and TODO'd it with a link back to this thread.